File: apps/solana/programs/contract/src/utils/discriminator.rs
use solana_program::hash::hash;

/// Computes the 8-byte Anchor-style instruction discriminator: `sha256("global:<name>")[0..8]`.
pub fn ix_disc(snake_name: &str) -> [u8; 8] {
    sha256_8(&format!("global:{snake_name}"))
}

/// Computes the 8-byte Anchor-style account discriminator: `sha256("account:<Name>")[0..8]`.
pub fn account_disc(pascal_name: &str) -> [u8; 8] {
    sha256_8(&format!("account:{pascal_name}"))
}

/// Computes the 8-byte Anchor-style event discriminator: `sha256("event:<Name>")[0..8]`.
pub fn event_disc(pascal_name: &str) -> [u8; 8] {
    sha256_8(&format!("event:{pascal_name}"))
}

fn sha256_8(preimage: &str) -> [u8; 8] {
    let mut out = [0u8; 8];
    out.copy_from_slice(&hash(preimage.as_bytes()).to_bytes()[..8]);
    out
}

File: packages/contract/src/discriminator.ts
import { createHash } from "node:crypto";

/**
 * Anchor-style discriminator helpers. The native-rust program at
 * `apps/solana/programs/contract` uses these exact preimages — see
 * `programs/contract/src/utils/discriminator.rs`. Keeping them in sync
 * is what lets a non-Anchor TS client talk to a non-Anchor Rust program
 * while remaining wire-compatible with anything else that speaks the
 * Anchor convention.
 */
function sha256_8(preimage: string): Buffer {
  return createHash("sha256").update(preimage).digest().subarray(0, 8);
}

/** Instruction discriminator: `sha256("global:<snake_name>")[0..8]`. */
export function ix_disc(snake_name: string): Buffer {
  return sha256_8(`global:${snake_name}`);
}

/** Account discriminator: `sha256("account:<PascalName>")[0..8]`. */
export function account_disc(pascal_name: string): Buffer {
  return sha256_8(`account:${pascal_name}`);
}

/** Event discriminator: `sha256("event:<PascalName>")[0..8]`. */
export function event_disc(pascal_name: string): Buffer {
  return sha256_8(`event:${pascal_name}`);
}
